Add `common_parent` relationship attribute

gmazoyer opened the pull request:
This relationship attribute is used to specify that peers of a given
relationship must share a common parent with the source object. While
the parent of the source object is infered automatically (as there is
only one parent per object), the `common_parent` value must be set to
the name of the parent relationship of the peer schema.

gmazoyer committed:
Add `common_parent` relationship attribute

This relationship attribute is used to specify that peers of a given
relationship must share a common parent with the source object. While
the parent of the source object is infered automatically (as there is
only one parent per object), the `common_parent` value must be set to
the name of the parent relationship of the peer schema.

gmazoyer committed:
Add schema validation when `common_parent` is set

gmazoyer committed:
Add peer parent relationship constraint validator

gmazoyer committed:
Support peer parent validation in `RelationshipAdd`

gmazoyer commented on backend/infrahub/core/relationship/constraints/peer_parent.py:
```diff
@@ -0,0 +1,54 @@
+        """Validate that all peers of a given `relm` have the same parent for the given `relationship_name`."""
+        node_parent = await node.get_parent_relationship_peer(db=self.db, name=parent_rel_name)
+        if not node_parent:
+            raise ValidationError(f"Node {node.id} ({node.get_kind()}) does not have a parent peer")
```
Considering the schema restrictions we are not expecting to fall into this, should we keep though? It also makes mypy happy.

ogenstad replied:
Yes if it makes mypy happy, but we could ad a comment to indicate that we don't expect to ever hit that point, it could make a future refactoring easier.

gmazoyer commented on backend/infrahub/core/relationship/constraints/peer_parent.py:
```diff
@@ -0,0 +1,54 @@
+        parents: set[str] = {node_parent.id}
+        for peer in peers.values():
+            parent = await peer.get_parent_relationship_peer(db=self.db, name=parent_rel_name)
+            if not parent:
+                raise ValidationError(f"Peer {peer.id} ({peer.get_kind()}) does not have a parent peer")
```
Same comment as above.

ogenstad replied:
Same comment as above :)

gmazoyer commented on backend/infrahub/graphql/mutations/relationship.py:
```diff
@@ -406,6 +407,35 @@ async def _validate_peer_types(
+    if len(parents) > 1:
+        raise ValidationError(
+            f"Cannot relate {source_node.id!r} to '{relationship_name}' peers that do not have the same parent"
+        )
```
Feels like duplicating logic from another place, but the constraint checker cannot run here for now.

ajtmccarty replied:
why can't the `RelationshipPeerParentConstraint` constraint run here?

gmazoyer replied:
Not sure yet, that the first thing I tried. I tried running it after during the transaction of the `RelationshipAdd` mutation but it failed like `neo4j.exceptions.ClientError: Explicit Transaction must be handled explicitly`. Running it before the transaction does not work because the checker isn't able to pick up the relationships are they are not set when it runs and therefore not picking up possible issues. Might need some more thoughts on this to figure it out.

gmazoyer committed:
Fix typo

gmazoyer committed:
Add function test

gmazoyer committed:
Really add test

ajtmccarty commented on backend/infrahub/core/schema/schema_branch.py:
```diff
@@ -975,6 +975,28 @@ def validate_names(self) -> None:
+        if not node.has_parent_relationship:
+            raise ValueError(
+                f"{node.kind}: Relationship {rel.name!r} defines 'common_parent' but node does not have a parent relationship"
+            ) from None
```
`from None` seems unnecessary

ajtmccarty commented on backend/infrahub/dependencies/builder/constraint/relationship_manager/peer_parent.py:
👍

gmazoyer committed:
Remove extra `from None` on exception raise

ogenstad commented on backend/infrahub/core/constraint/node/runner.py:
```diff
@@ -38,4 +38,6 @@ async def check(
                 for relationship_constraint in self.relationship_manager_constraints:
-                    await relationship_constraint.check(relm=relationship_manager, node_schema=node.get_schema())
+                    await relationship_constraint.check(
+                        relm=relationship_manager, node_schema=node.get_schema(), node=node
```
As we're adding the node as an argument here and we can access the schema with node.get_schema() the node_schema parameter gets a bit redundant now. But we can perhaps consider removing that as a cleanup step.

gmazoyer replied:
I will do that in a dedicated PR as it needs to change some code in other constraint checkers.

gmazoyer committed:
Add some comments

gmazoyer committed:
Set `common_*` rel attribute update support to allowed

They will be left like that until we have proper constraint validators
in place.
